### ISSUE TITLE
RuntimeError "There is no current event loop in thread" if not on main Thread

### ISSUE THREAD

**Issue by redreceipt** [graphql-python/gql#134]
I'm getting an error using the basic example syntax. You mention a warning below on the README and if this is it, I apologize, but I'm not familiar enough with AsyncIO to connect the dots. 

My code:
```py
class GraphQLClient:

    transport = AIOHTTPTransport(url="https://fake.com/graphql")
    client = Client(transport=transport, fetch_schema_from_transport=True)

    def request(self, body):
        return self.client.execute(body)

client = GraphQLClient()
r = client.request(
    gql(
    """
    query getUserID {
      login( email_or_phone_or_username: "{username}" password:"{pw}") {
        user_id
      }
    }
    """
    )
 )

```

The error:
```
  File "/Users/michael/Documents/Projects/kpffl/sleeper.py", line 16, in request
    return self.client.execute(body)
  File "/Users/michael/Documents/Projects/kpffl/venv/lib/python3.8/site-packages/gql/client.py", line 113, in execute
    loop = asyncio.get_event_loop()
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/events.py", line 639, in get_event_loop
    raise RuntimeError('There is no current event loop in thread %r.'
RuntimeError: There is no current event loop in thread 'Thread-8'.
```

**Comment by leszekhanusz** [comment]
> You mention a warning below on the README and if this is it

No, it is something else. Your code is working for me.

Are you on a Mac ?
Are you using this code inside a new Thread ?

**Comment by leszekhanusz** [comment]
Could you try to add before:

```python
import asyncio
asyncio.set_event_loop(asyncio.new_event_loop())
```

**Comment by redreceipt** [comment]
That may have worked. Now I'm getting a different error:

`TypeError: Invalid or incomplete introspection result. Ensure that you are passing the 'data' attribute of an introspection response and no 'errors' were returned alongside: None.`

I'm running inside a Flask instance which I believe is multi-threaded, could that be why I needed to add the asyncio line?

**Comment by leszekhanusz** [comment]
> I'm running inside a Flask instance which I believe is multi-threaded, could that be why I needed to add the asyncio line?

Yes, we should fix this.

For your next error, it means that the introspection query to fetch the schema failed. I will need more info.
Could you please try fetch_schema_from_transport=False ?

**Comment by redreceipt** [comment]
I modified slightly with this code:

```py
class GraphQLClient:
    def __init__(self):
        asyncio.set_event_loop(asyncio.new_event_loop())
        transport = AIOHTTPTransport(url="https://cool.api/graphql")
        self.client = Client(transport=transport, fetch_schema_from_transport=True)
        print(self.client.introspection)

    def request(self, body):
        return self.client.execute(body)
```

but `client.introspection` comes back `None`

**Comment by redreceipt** [comment]
> EDIT: which backend are you using?

Flask

**Comment by leszekhanusz** [comment]
The schema is not fetched directly with async transports but only when doing the first execute. Could you try something like this:

```python
    def request(self, body):
        try:
            result = self.client.execute(body)
        except:
            print(self.client.introspection)

        return result
```

**Comment by leszekhanusz** [comment]
Also Flask is a standard web server, it does not support GraphQL by default. Do you use something else like `Flask-GraphQL`.
Or did you implement all the backend yourself directly using Flask ?

**Comment by leszekhanusz** [comment]
And what result do you have with `fetch_schema_from_transport=False`:

```python
client = Client(transport=transport, fetch_schema_from_transport=False)
```

**Comment by redreceipt** [comment]
> The schema is not fetched directly with async transports but only when doing the first execute. Could you try something like this:
> 
> ```python
>     def request(self, body):
>         try:
>             result = self.client.execute(body)
>         except:
>             print(self.client.introspection)
> 
>         return result
> ```

```py
    def request(self, body, variables=None):
        try:
            return self.client.execute(body, variable_values=variables)
        except:
            print(self.client.introspection)
```

still shows `self.client.introspection` being `None`

**Comment by redreceipt** [comment]
> Also Flask is a standard web server, it does not support GraphQL by default. Do you use something else like `Flask-GraphQL`.
> Or did you implement all the backend yourself directly using Flask ?

It's not my GraphQL server, it's a functioning third-party GQL server. It's just making calls from my Flask site acting as the client to their server.

**Comment by leszekhanusz** [comment]
Sorry my question about the backend was about the GraphQL backend, do you know which type of backend is used on this GraphQL server ?

**Comment by redreceipt** [comment]
> And what result do you have with `fetch_schema_from_transport=False`:
> 
> ```python
> client = Client(transport=transport, fetch_schema_from_transport=False)
> ```

This seems to work for me! So they must just have introspection turned off. It's fine, I don't really need it. So maybe your lib could handle the error a little more gracefully? Somehow let me know that introspection is a no go?

**Comment by redreceipt** [comment]
> Sorry my question about the backend was about the GraphQL backend, do you know which type of backend is used on this GraphQL server ?

Oh no I'm sorry I don't.

**Comment by redreceipt** [comment]
If you're exploring how to handle it though, Node.js Apollo lets you turn introspection off. I could make a reasonable guess that is the server that is being used and you could build an error handling solution against that?

**Comment by leszekhanusz** [comment]
> So maybe your lib could handle the error a little more gracefully?

Yes, probably. I think we could just put a warning with the result of the introspection query if it is not parsed correctly and continue without a local schema. It will just disable local schema validation so the queries should still work.

**Comment by leszekhanusz** [comment]
Note: could you please try the PR #135 and verify that you don't need the `asyncio.set_event_loop(asyncio.new_event_loop())` line anymore?

**Comment by redreceipt** [comment]
That was fast! I'd love to test it. I'm a rookie though, how could I pull the code from that PR to test?

**Comment by leszekhanusz** [comment]
It is explained in the [CONTRIBUTING file](https://github.com/graphql-python/gql/blob/master/CONTRIBUTING.md)

```bash
# Clone my repo somewhere
git clone https://github.com/leszekhanusz/gql.git
cd gql
# go to the correct branch
git checkout fix_get_current_loop_in_new_thread

# Install conda if you don't have it already ...

# Then create a conda environment, and install the dependencies inside:
conda create -n gql-dev python=3.8
conda activate gql-dev
pip install -e.[dev]
```

**Comment by redreceipt** [comment]
Tested and confirmed ✅

Actually found a much easier way on Google:
`pip install git+https://github.com/leszekhanusz/gql.git@fix_get_current_loop_in_new_thread`

**Comment by leszekhanusz** [comment]
> Actually found a much easier way on Google

That will be useful for me!

**Comment by leszekhanusz** [comment]
This is now fixed in version 3.0.0a2
